### PR TITLE
Initial preprocessor work; untested

### PR DESCRIPTION
**Pull request by rabbitfang**
Initial work on preprocessor. None of this has been tested, but gives us a good starting point. In the `utils` directory, `ref.py`, `visit.py` and `yaml.py` were pulled from the linker prototype.

### CODE REVIEW

File: opta/utils/ref.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
+import functools
+import re
+from collections import abc
+from typing import Any, Iterable, Iterator, Sequence, Tuple, TypeVar, Union, overload
+
+from opta.utils.yaml import register_yaml_class
+
+_PART_REGEX = re.compile(r"[a-z]([a-z0-9_]*[a-z0-9])?", re.IGNORECASE)
```
Comment (bigbitbus) [inline opta/utils/ref.py:8]: Is there a reason why a string ending with a underscore is not acceptable here?

**Review by bigbitbus — APPROVED** [review]
Great start! Loved the way you brought in multiple test cases per test with the pytest.mark.parameterize.

I added a few questions in the code; we can discuss them on a call or you can just put in comments for everyone.

File: opta/preprocessor/util.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+DEFAULT_VERSION = 1
+CURRENT_VERSION = 2
+
+Handler = Callable[[dict], None]
+
+
+class VersionProcessor:
+    """
+    Processes a layer YAML for specific versions.
```
[the remark this answers is not in the document]
Comment (rabbitfang) [reply opta/preprocessor/util.py:13]: While we removed versioning from the module API scope, I needed a way to figure out if we needed to transform the YAML to the "new" format or if it was already there, and I couldn't think of a better way of doing that than versioning the YAML.

File: opta/preprocessor/v1.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            continue
+
+        if format_spec:
+            raise ValueError("format_spec in formatted string not supported")
+
+        if conversion:
+            raise ValueError("conversion field in formatted string not supported")
+
+        try:
+            Reference.parse(field_name)
```
[the remark this answers is not in the document]
Comment (rabbitfang) [reply opta/preprocessor/v1.py:49]: This is only verifying that `field_name` is a valid reference. If we ignored issues now, we would just have them later when we try to parse the references again.

File: opta/utils/ref.py (hunk 1 above)
Comment (rabbitfang) [reply opta/utils/ref.py:8]: Not really, but there probably shouldn't be any keys ending in underscores. We can allow it if we want.

File: tests/preprocessor/test_v1.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        assert output == expected
+
+    @pytest.mark.parametrize(
+        "input,expected",
+        [
+            ["foo", "foo"],
+            ["foo {bar}", "foo ${bar}"],
+            ["{spam} {ham}", "${spam} ${ham}"],
+            ["${foo}", "$${foo}"],
+        ],
```
[the remark this answers is not in the document]
Comment (rabbitfang) [reply tests/preprocessor/test_v1.py:35]: I think in the notion doc it was just an example showing the formatting difference. An empty reference wouldn't be valid anyways.

**Pull request merged**